moshi/src/main/java/com/squareup/moshi/ArrayJsonAdapter.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (C) 2014 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi
+
+import kotlin.Throws
+import java.lang.reflect.Array
+import java.lang.reflect.Type
+import okio.IOException
+
+/**
+ * Converts arrays to JSON arrays containing their converted contents. This supports both primitive
+ * and object arrays.
+ */
+internal class ArrayJsonAdapter(private val elementClass: Class<*>, private val elementAdapter: JsonAdapter<Any>
+) : JsonAdapter<Any?>() {
+  @Throws(IOException::class)
+  override fun fromJson(reader: JsonReader): Any {
+    val list = mutableListOf<Any?>()
+    reader.beginArray()
+    while (reader.hasNext()) {
+      list.add(elementAdapter.fromJson(reader))
+    }
+    reader.endArray()
+    val array = Array.newInstance(elementClass, list.size)
+    list.forEachIndexed { i, item ->
+      Array.set(array, i, item)
+    }
+    return array
+  }
+
+  @Throws(IOException::class)
+  override fun toJson(writer: JsonWriter, value: Any?) {
+    writer.beginArray()
+    var i = 0
+    val size = Array.getLength(value)
```

Comment:
Same here - can we use idiomatic Kotlin APIs?

Reply:
I changed the loop style, is there a more idiomatic way to get the size?

Reply:
cast `value` to `Array<Any>` first? This will throw if it's not an array, better to just do it ourselves rather than use these reflection APIs